# <font color=blue>Assignments for "Evaluating Goodness of Fit"</font>

As in previous lessons, please submit a link to a single gist that contains links to two Juypyter notebooks (one for each assignment below).

## 1. Weather model

For this assignment, you'll revisit the historical temperature dataset. To complete this assignment, submit a link in the gist file to the Jupyter notebook containing your solutions to the following tasks:

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.api as sm
weather_df = pd.read_csv("https://djl-lms-assets.s3.eu-central-1.amazonaws.com/datasets/weatherHistory.csv")
weather_df.head()

Formatted Date        Summary Precip Type  Temperature (C)  \
0  2006-04-01 00:00:00.000 +0200  Partly Cloudy        rain         9.472222   
1  2006-04-01 01:00:00.000 +0200  Partly Cloudy        rain         9.355556   
2  2006-04-01 02:00:00.000 +0200  Mostly Cloudy        rain         9.377778   
3  2006-04-01 03:00:00.000 +0200  Partly Cloudy        rain         8.288889   
4  2006-04-01 04:00:00.000 +0200  Mostly Cloudy        rain         8.755556   

   Apparent Temperature (C)  Humidity  Wind Speed (km/h)  \
0                  7.388889      0.89            14.1197   
1                  7.227778      0.86            14.2646   
2                  9.377778      0.89             3.9284   
3                  5.944444      0.83            14.1036   
4                  6.977778      0.83            11.0446   

   Wind Bearing (degrees)  Visibility (km)  Loud Cover  Pressure (millibars)  \
0                   251.0          15.8263         0.0               1015.13   
1                   259.0          15.8263         0.0               1015.63   
2                   204.0          14.9569         0.0               1015.94   
3                   269.0          15.8263         0.0               1016.41   
4                   259.0          15.8263         0.0               1016.51   

                       Daily Summary  
0  Partly cloudy throughout the day.  
1  Partly cloudy throughout the day.  
2  Partly cloudy throughout the day.  
3  Partly cloudy throughout the day.  
4  Partly cloudy throughout the day.

- Like in the previous lesson, build a linear regression model where your target variable is the difference between the *apparenttemperature* and the *temperature*. As explanatory variables, use *humidity* and *windspeed*. Now, estimate your model using OLS. What are the R-squared and adjusted R-squared values? Do you think they are satisfactory? Why? 

In [2]:
weather_df.insert(3, "Temperature Difference", 
               (weather_df["Temperature (C)"] - weather_df["Apparent Temperature (C)"]))

In [3]:
X = weather_df[["Humidity", "Wind Speed (km/h)"]]
y = weather_df["Temperature Difference"]

X = sm.add_constant(X)
results = sm.OLS(y, X).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                              OLS Regression Results                              
==================================================================================
Dep. Variable:     Temperature Difference   R-squared:                       0.288
Model:                                OLS   Adj. R-squared:                  0.288
Method:                     Least Squares   F-statistic:                 1.949e+04
Date:                    Sun, 09 May 2021   Prob (F-statistic):               0.00
Time:                            17:04:40   Log-Likelihood:            -1.7046e+05
No. Observations:                   96453   AIC:                         3.409e+05
Df Residuals:                       96450   BIC:                         3.409e+05
Df Model:                               2                                         
Covariance Type:                nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                -2.4381      0.021   -115.948      0.000      -2.479      -2.397
Humidity              3.0292      0.024    126.479      0.000       2.982       3.076
Wind Speed (km/h)     0.1193      0.001    176.164      0.000       0.118       0.121
==============================================================================
Omnibus:                     3935.747   Durbin-Watson:                   0.264
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4613.311
Skew:                           0.478   Prob(JB):                         0.00
Kurtosis:                       3.484   Cond. No.                         88.1
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**R-squared = Adj.R-squared = 0.288 Which is very low. In brief, our model %28 of the variance for a dependent variable that's explained by an independent variable or variables.**

- Next, include the interaction of *humidity* and *windspeed* to the model above and estimate the model using OLS. Now, what is the R-squared of this model? Does this model improve upon the previous one? 


In [4]:
weather_df.insert(5, "Humidity*Wind Speed", 
               (weather_df.Humidity * weather_df["Wind Speed (km/h)"]))

weather_df["Humidity x Wind Speed"] = weather_df.Humidity * weather_df["Wind Speed (km/h)"]
XX = weather_df[["Humidity", "Wind Speed (km/h)","Humidity*Wind Speed"]]
y = weather_df["Temperature Difference"]

XX = sm.add_constant(XX)
results_new = sm.OLS(y, XX).fit()
results_new.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                              OLS Regression Results                              
==================================================================================
Dep. Variable:     Temperature Difference   R-squared:                       0.341
Model:                                OLS   Adj. R-squared:                  0.341
Method:                     Least Squares   F-statistic:                 1.666e+04
Date:                    Sun, 09 May 2021   Prob (F-statistic):               0.00
Time:                            17:04:40   Log-Likelihood:            -1.6669e+05
No. Observations:                   96453   AIC:                         3.334e+05
Df Residuals:                       96449   BIC:                         3.334e+05
Df Model:                               3                                         
Covariance Type:                nonrobust                                         
=======================================================================================
                          coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------
const                  -0.0839      0.033     -2.511      0.012      -0.149      -0.018
Humidity               -0.1775      0.043     -4.133      0.000      -0.262      -0.093
Wind Speed (km/h)      -0.0905      0.002    -36.797      0.000      -0.095      -0.086
Humidity*Wind Speed     0.2971      0.003     88.470      0.000       0.291       0.304
==============================================================================
Omnibus:                     4849.937   Durbin-Watson:                   0.262
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             9295.404
Skew:                           0.378   Prob(JB):                         0.00
Kurtosis:                       4.320   Cond. No.                         193.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**R-squared increased to 0.341. Results can explain %34 of the variance for a temperature difference. Which is still low.**

- Add *visibility* as additional explanatory variable to the first model and estimate it. Did R-squared increase? What about adjusted R-squared? Compare the differences put on the table by the interaction term and the *visibility* in terms of the improvement in the adjusted R-squared. Which one is more useful?

In [7]:
XXX = weather_df[["Humidity", "Wind Speed (km/h)", "Visibility (km)"]]

x13 = sm.add_constant(XXX)
results_new2 = sm.OLS(y, XXX).fit()
results_new2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                   OLS Regression Results                                  
===========================================================================================
Dep. Variable:     Temperature Difference   R-squared (uncentered):                   0.490
Model:                                OLS   Adj. R-squared (uncentered):              0.490
Method:                     Least Squares   F-statistic:                          3.094e+04
Date:                    Sun, 09 May 2021   Prob (F-statistic):                        0.00
Time:                            17:23:27   Log-Likelihood:                     -1.7096e+05
No. Observations:                   96453   AIC:                                  3.419e+05
Df Residuals:                       96450   BIC:                                  3.419e+05
Df Model:                               3                                                  
Covariance Type:                nonrobust                                                  
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
Humidity              1.3488      0.012    108.590      0.000       1.324       1.373
Wind Speed (km/h)     0.1052      0.001    167.634      0.000       0.104       0.106
Visibility (km)      -0.0976      0.001   -110.936      0.000      -0.099      -0.096
==============================================================================
Omnibus:                     5476.521   Durbin-Watson:                   0.280
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             6619.177
Skew:                           0.587   Prob(JB):                         0.00
Kurtosis:                       3.519   Cond. No.                         43.9
==============================================================================

Notes:
[1] R² is computed without centering (uncentered) since the model does not contain a constant.
[2] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

- Choose the best one from the three models above with respect to their AIC and BIC scores. Validate your choice by discussing your justification with your mentor.

**Last model is better than others. Because;**
- R-Square and F-Statistic got the highest scores.
- AIC and BIC scores have the highest scores.

##  2. House prices model

In this exercise, you'll work on your house prices model. To complete this assignment, submit a link in the gist file to the Jupyter notebook containing your solutions to the following tasks:

In [5]:
house_prices_df = pd.read_csv("https://djl-lms-assets.s3.eu-central-1.amazonaws.com/datasets/house_prices.csv", sep = ";")
house_prices_df.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

- Run your house prices model again and assess the goodness of fit of your model using F-test, R-squared, adjusted R-squared, AIC and BIC.

In [8]:
y1 = house_prices_df["SalePrice"]
X1 = house_prices_df[['OverallQual', 'TotalBsmtSF','1stFlrSF','GrLivArea','GarageCars','GarageArea','YrSold']]

X1 = sm.add_constant(X1)
results1 = sm.OLS(y1, X1).fit()
results1.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.762
Model:                            OLS   Adj. R-squared:                  0.761
Method:                 Least Squares   F-statistic:                     663.8
Date:                Sun, 09 May 2021   Prob (F-statistic):               0.00
Time:                        17:30:32   Log-Likelihood:                -17496.
No. Observations:                1460   AIC:                         3.501e+04
Df Residuals:                    1452   BIC:                         3.505e+04
Df Model:                           7                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const       -3.642e+05   1.54e+06     -0.237      0.813   -3.38e+06    2.66e+06
OverallQual    2.4e+04   1083.767     22.141      0.000    2.19e+04    2.61e+04
TotalBsmtSF    24.3949      4.319      5.648      0.000      15.922      32.867
1stFlrSF       11.1760      5.034      2.220      0.027       1.301      21.051
GrLivArea      43.1346      2.681     16.089      0.000      37.875      48.394
GarageCars   1.453e+04   3020.839      4.810      0.000    8604.032    2.05e+04
GarageArea     15.6356     10.480      1.492      0.136      -4.922      36.193
YrSold        130.2681    766.892      0.170      0.865   -1374.067    1634.603
==============================================================================
Omnibus:                      431.830   Durbin-Watson:                   1.976
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            39204.109
Skew:                          -0.314   Prob(JB):                         0.00
Kurtosis:                      28.378   Cond. No.                     4.61e+06
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 4.61e+06. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

- Do you think your model is satisfactory? If so, why?

**YrSold t-test score is higher than 0,05 which means it has not an impact on our model. It must be removed.**

**R-squared value between 70-90% is ideal. In the model is 76% which is ok**

- In order to improve the goodness of fit of your model, try different model specifications by adding or removing some variables. 

In [10]:
X1 = house_prices_df[['OverallQual', 'TotalBsmtSF','1stFlrSF','GrLivArea','GarageCars','GarageArea']]

X1 = sm.add_constant(X1)
results11 = sm.OLS(y1, X1).fit()
results11.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.762
Model:                            OLS   Adj. R-squared:                  0.761
Method:                 Least Squares   F-statistic:                     775.0
Date:                Sun, 09 May 2021   Prob (F-statistic):               0.00
Time:                        17:37:00   Log-Likelihood:                -17496.
No. Observations:                1460   AIC:                         3.501e+04
Df Residuals:                    1453   BIC:                         3.504e+04
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const       -1.027e+05   4903.994    -20.932      0.000   -1.12e+05    -9.3e+04
OverallQual    2.4e+04   1083.393     22.150      0.000    2.19e+04    2.61e+04
TotalBsmtSF    24.3908      4.318      5.649      0.000      15.921      32.860
1stFlrSF       11.1859      5.032      2.223      0.026       1.315      21.057
GrLivArea      43.1229      2.679     16.095      0.000      37.867      48.379
GarageCars   1.452e+04   3018.621      4.809      0.000    8593.872    2.04e+04
GarageArea     15.6639     10.475      1.495      0.135      -4.884      36.212
==============================================================================
Omnibus:                      431.781   Durbin-Watson:                   1.976
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            39208.253
Skew:                          -0.313   Prob(JB):                         0.00
Kurtosis:                      28.380   Cond. No.                     1.14e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.14e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [16]:
X11 = house_prices_df[['OverallQual', 'TotalBsmtSF','1stFlrSF','GrLivArea','GarageCars','GarageArea','YearBuilt']]

X11 = sm.add_constant(X11)
results111 = sm.OLS(y1, X11).fit()
results111.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.770
Model:                            OLS   Adj. R-squared:                  0.769
Method:                 Least Squares   F-statistic:                     693.0
Date:                Sun, 09 May 2021   Prob (F-statistic):               0.00
Time:                        19:31:15   Log-Likelihood:                -17472.
No. Observations:                1460   AIC:                         3.496e+04
Df Residuals:                    1452   BIC:                         3.500e+04
Df Model:                           7                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const       -6.942e+05   8.49e+04     -8.174      0.000   -8.61e+05   -5.28e+05
OverallQual  2.079e+04   1160.683     17.916      0.000    1.85e+04    2.31e+04
TotalBsmtSF    20.2475      4.290      4.720      0.000      11.833      28.662
1stFlrSF       13.4436      4.962      2.709      0.007       3.710      23.177
GrLivArea      48.0007      2.727     17.599      0.000      42.650      53.351
GarageCars   9969.4283   3040.904      3.278      0.001    4004.393    1.59e+04
GarageArea     16.8085     10.309      1.631      0.103      -3.413      37.030
YearBuilt     310.9118     44.563      6.977      0.000     223.496     398.327
==============================================================================
Omnibus:                      449.404   Durbin-Watson:                   1.981
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            48372.513
Skew:                          -0.326   Prob(JB):                         0.00
Kurtosis:                      31.191   Cond. No.                     2.57e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.57e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

- For each model you try, get the goodness of fit metrics and compare your models with each other. Which model is the best and why?

**1st model:**
- R-squared:0.762  Adj.R-squared:0.761
- F-statistic:663.8
- AIC:3.501  BIC:3.505

**2nd model**
- R-squared:0.762  Adj.R-squared:0.761
- F-statistic:775.0
- AIC: 3.501  BIC:3.504

**3rd model**
- R-squared:0.769 Adj.R-squared:0.768
- F-statistic:	807.2
- AIC:3.496  BIC:3.500	

**According to R-squared, Adj.R-squared and F-statistic 3rd model has the best performance.**

**On the other hand, 1st modal has better AIC and BIC scores.**